### PR TITLE
bpo-44133: Export Py_FrozenMain() symbol.

### PR DESCRIPTION
**Pull request by shihai1991**
The Python binary now builds the libpython static library using
"-Wl,--whole-archive" or "-Wl,-all_load"linker option to
export all symbols exported by object files.
Previously, the "Py_FrozenMain" symbol was not exported.

Co-authored-by: Victor Stinner <vstinner@python.org>

https://bugs.python.org/issue44133

### CODE REVIEW

**Comment by shihai1991** [comment]
I tested this PR in my centos vm and it has passed.
Due to lack of the MacOS, I haven't test the `-all_load` link option for lacking of macos.

File: configure [hunk 1]
```diff
@@ -1509,7 +1498,7 @@ fi
 for ac_var in	exec_prefix prefix bindir sbindir libexecdir datarootdir \
 		datadir sysconfdir sharedstatedir localstatedir includedir \
 		oldincludedir docdir infodir htmldir dvidir pdfdir psdir \
-		libdir localedir mandir runstatedir
```
Comment (tiran) [inline configure]: You seem to use a patched or modified version of autoconf 2.69. Could you please use an unpatched version to reduce noise? You can use my container image https://quay.io/repository/tiran/cpython_autoconf

File: configure [hunk 2]
```diff
@@ -6481,8 +6469,14 @@ else # shared is disabled
...
-          BLDLIBRARY='$(LIBRARY)'
-          LDLIBRARY='libpython$(LDVERSION).dll.a'
+      BLDLIBRARY='$(LIBRARY)'
+      LDLIBRARY='libpython$(LDVERSION).dll.a'
+          ;;
+    Linux*|GNU*|NetBSD*|FreeBSD*|DragonFly*|OpenBSD*|VxWorks*)
+	  BLDLIBRARY='-Wl,--whole-archive $(LDLIBRARY) -Wl,--no-whole-archive'
+          ;;
+    Darwin*)
+	  BLDLIBRARY='-Wl,-all_load $(LDLIBRARY)'
```
Comment (tiran) [inline configure]: You are mixing tabs and spaces. Could you please use spaces-only and ``AS_CASE``?

```suggestion
  AS_CASE([$ac_sys_system],
    [CYGWIN*], [
      BLDLIBRARY='$(LIBRARY)'
      LDLIBRARY='libpython$(LDVERSION).dll.a'
    ], 
    [Linux*|GNU*|NetBSD*|FreeBSD*|DragonFly*|OpenBSD*|VxWorks*], [
      BLDLIBRARY='-Wl,--whole-archive $(LDLIBRARY) -Wl,--no-whole-archive'
    ],
    [Darwin*], [BLDLIBRARY='-Wl,-all_load $(LDLIBRARY)']
  )
```

**Review by tiran — CHANGES_REQUESTED** [review]

File: configure (hunk 2 above)
Comment (shihai1991) [reply configure:6479]: > You are mixing tabs and spaces.
    I never use tabs in my local in fact. I am confused too. this patch's configure file in my local.
    Let me check it again.
image

File: configure (hunk 1 above)
Comment (shihai1991) [reply configure]: OK, thanks. Let me have a try.

File: configure (hunk 2 above)
Comment (shihai1991) [reply configure]: OK, Let me check it again.

File: configure (hunk 1 above)
Comment (shihai1991) [reply configure]: Thanks, your docker image works :)

File: configure (hunk 2 above)
Comment (shihai1991) [reply configure]: OK, I forget my automative tabs when I press the `enter`.
Updated, thanks.

**Comment by vstinner** [comment]
I suggest to not use libpython.a to build the python executable, but pass directly the long list of object filenames (.o).

**Comment by shihai1991** [comment]
> I suggest to not use libpython.a to build the python executable, but pass directly the long list of object filenames (.o).

OK, it's would be more direct.

**Comment by vstinner** [comment]
I wrote PR 30556 to propose my idea.

**Comment by vstinner** [comment]
The problem of this approach is that the ``-Wl,--whole-archive`` flag is not portable. That's why you had to use ``-Wl,-all_load`` for example.

I wrote a different approach which simply skips the static Python library: GH-30556.

**Pull request closed without merging**